### PR TITLE
[VPD-1087]: Phase2 - Token Converter

### PR DESCRIPTION
**Pull request by Debugger022**
## Summary

Replaces the 5-contract Token Converter system (RiskFundConverter + 4 Prime converters + XVSVaultConverter + WBNBBurnConverter + ConverterNetwork) with a single `TokenBuyback` contract class driven by an ACM-authorized finance-team cron, eliminating the 50% community premium, oracle dependency, and idle-balance problem.

On BSC: **8 legacy contracts → 7 TokenBuyback proxies** (4 Prime + RiskFund + XVS + Treasury), with `WBNBBurnConverter` and `ConverterNetwork` retired.

### `TokenBuyback` (new, ~290 lines)
- Immutables: `DESTINATION`, `BASE_ASSET`, `IS_RISK_FUND`, `PROTOCOL_SHARE_RESERVE` — set in constructor, no storage slot.
- `executeBuyback(tokenIn, amountIn, minAmountOut, deadline, router, routerCalldata, comptroller)` — ACM-gated. Router allowlist, deadline / slippage / comptroller / InvalidTokenIn / InsufficientBalance guards. Balance-diff measured at `address(this)` to block donation inflation. Post-swap `safeTransfer` to `DESTINATION`, conditional `RiskFundV2.updatePoolState` when `IS_RISK_FUND`.
- `forwardBaseAsset(comptroller, amount)` — ACM-gated. Caller-specified partition so each BASE_ASSET contributor is credited its own share across the downstream per-pool ledger.
- `updateAssetsState(comptroller, asset)` — gated to `PROTOCOL_SHARE_RESERVE` only. Emits per-delivery delta (watermark pattern via `assetsReserves[asset]`), resynced after every outflow so inflow deltas stay correct across interleaved deposits / withdrawals. Prevents spoofed AssetsReceived events that could mis-attribute inflows to arbitrary comptrollers.
- `setAllowedRouter`, `sweepToken` — `onlyOwner` (Timelock).
- Uses `AddressUpgradeable.functionCall` to propagate router errors.

### `RiskFundV2` (upgrade)
- `updatePoolState`: replaced hardcoded `msg.sender == riskFundConverter` with ACM permission check — one implementation serves any number of buyback callers.
- `preSweepToken`: simplified to a plain balance check (no longer iterates `getPools()` on the retired converter).
- `IRiskFundConverter` import and `InvalidRiskFundConverter` error removed.

### `ITokenBuyback` (new interface)
- Extends `IIncomeDestination` with `executeBuyback`, `forwardBaseAsset`, `setAllowedRouter`, `sweepToken`.

## Deploy scripts (BSC mainnet + testnet)

| File | Instances | BASE_ASSET | DESTINATION | IS_RISK_FUND |
|---|---|---|---|---|
| [deploy/006-risk-fund-buyback.ts](deploy/006-risk-fund-buyback.ts) | `RiskFundBuyback` | USDT | RiskFundV2 | true |
| [deploy/007-prime-buyback.ts](deploy/007-prime-buyback.ts) | `USDTPrimeBuyback`, `USDCPrimeBuyback`, `BTCBPrimeBuyback`, `ETHPrimeBuyback` | USDT / USDC / BTCB / ETH | PrimeLiquidityProvider | false |
| [deploy/008-xvs-buyback.ts](deploy/008-xvs-buyback.ts) | `XVSBuyback` | XVS | XVSVaultTreasury | false |
| [deploy/009-treasury-buyback.ts](deploy/009-treasury-buyback.ts) | `TreasuryBuyback` | USDT (placeholder — tokenomics TBD) | VTreasury | false |

Pattern: `OptimizedTransparentUpgradeableProxy` via `DefaultProxyAdmin`, initializer takes ACM only, ownership transferred to `NormalTimelock` after deploy (two-step, accept in VIP). New helper `helpers/deploymentConfig.ts::getContractAddressOrNullAddress`.

## Tests

- [tests/ProtocolReserve/TokenBuyback.ts](tests/ProtocolReserve/TokenBuyback.ts): **45 passing** — initialization, `updateAssetsState` delta regression + PSR gate + donation-spoof revert, `setAllowedRouter`, `sweepToken`, `executeBuyback` happy + revert paths, `forwardBaseAsset` per-pool partitioning + revert paths, donation-inflation regression.
- [tests/fork/TokenBuyback.ts](tests/fork/TokenBuyback.ts): **31 passing** against bscmainnet fork at block `93647884` — 5 E2E happy paths per instance (RiskFund / 4 Prime variants / XVS / Treasury), PSR-gate revert, donation-spoof block, output-redirect balance-diff guard, donation-inflation protection, sequential buybacks, full admin + edge case coverage.
- [contracts/Test/Mocks/MockRouter.sol](contracts/Test/Mocks/MockRouter.sol): DEX mock with `shouldFail` and `skipTransfer` toggles.

## Test plan

- [ ] `yarn compile` — clean
- [ ] `yarn lint` — no errors on changed files
- [ ] `yarn hardhat test tests/ProtocolReserve/TokenBuyback.ts` — 45 passing
- [ ] `yarn hardhat test tests/ProtocolReserve/RiskFundV2.ts` — ACM-gated path passing
- [ ] `FORK=true FORKED_NETWORK=bscmainnet ARCHIVE_NODE_bscmainnet=<archival> yarn hardhat test tests/fork/TokenBuyback.ts` — 31 passing
- [ ] `npx hardhat deploy --tags RiskFundBuyback,PrimeBuyback,XVSBuyback,TreasuryBuyback --network bsctestnet` — dry-run once bsctestnet external deployments wired
- [ ] Audit sign-off on the PSR gate + delta emission + forwardBaseAsset partitioning

## Open items
- `TreasuryBuyback` BASE_ASSET (`USDT` placeholder) pending tokenomics confirmation — immutable, so redeploy required if changed.
- `bsctestnet` external deployments in `hardhat.config.ts` do not currently include `@venusprotocol/venus-protocol` — deploy-script dry-run on testnet blocked until fixed (pre-existing infra issue).

### CODE REVIEW

File: contracts/ProtocolReserve/TokenBuyback.sol [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+    using AddressUpgradeable for address;
+
+    /// @notice Address where buyback proceeds are sent
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    address public immutable DESTINATION;
+
+    /// @notice Token that every swap converts into (output of every buyback)
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    address public immutable BASE_ASSET;
```
Comment (fred-venus) [inline contracts/ProtocolReserve/TokenBuyback.sol:30]: is it better to name it OUTPUT_ASSET, base asset sounds like input asset 😂

File: contracts/ProtocolReserve/TokenBuyback.sol [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+    }
+
+    /// @dev Transfers swap output to DESTINATION and notifies RiskFund if applicable
+    function _settleBuyback(uint256 amountOut, address comptroller) internal {
+        if (amountOut == 0) {
+            return;
+        }
+        IERC20Upgradeable(BASE_ASSET).safeTransfer(DESTINATION, amountOut);
+        if (IS_RISK_FUND) {
+            IRiskFund(DESTINATION).updatePoolState(comptroller, BASE_ASSET, amountOut);
```
Comment (fred-venus) [inline contracts/ProtocolReserve/TokenBuyback.sol]: Just checked, if it's to primeLiquidityProvider or XVSVaultTreasury, no callback is needed. But still curious why we need this for riskFund ? is it just for accounting purpose i.e. so that we know which pool contributes how many fund etc ?

File: contracts/ProtocolReserve/RiskFundV2.sol [hunk 3]
```diff
@@ -218,45 +211,8 @@ contract RiskFundV2 is AccessControlledV8, RiskFundV2Storage, IRiskFund {
     /// @param tokenAddress Address of the token
     /// @param amount Amount transferred to address(to)
     /// @custom:error InsufficientBalance is thrown when amount entered is greater than balance
-    function preSweepToken(address tokenAddress, uint256 amount) internal {
+    function preSweepToken(address tokenAddress, uint256 amount) internal view {
         uint256 balance = IERC20Upgradeable(tokenAddress).balanceOf(address(this));
         if (amount > balance) revert InsufficientBalance();
-
```
Comment (fred-venus) [inline contracts/ProtocolReserve/RiskFundV2.sol:224]: About RiskFund, what i am thinking:

1. non-bsc chain they dont have RiskFund
2. for bsc chain, isolated pool has been deprecated, the only scenario we use RickFund is for core pool

Based on above 2 conditions, does it still make sense to maintain `poolAssetsFunds[comptroller][asset]`, like we dont need per pool accounting anymore ? 


Also here u have already deleted the part i.e. pro-rata update of poolAssetsFunds, i am thinking should we remove everything related to poolAssetsFunds ? even sweepTokenFromPool, transferReserveForAuction, shortFallContract ...

Comment (fred-venus) [reply contracts/ProtocolReserve/RiskFundV2.sol:224]: But ofc, we still need to be compatible with existing deprecated isolated pool contract, there might still be some income, its just the per-controller accounting part is no longer necessary

File: contracts/ProtocolReserve/TokenBuyback.sol (hunk 1 above)
Comment (Debugger022) [reply contracts/ProtocolReserve/TokenBuyback.sol:30]: Kept `BASE_ASSET` for consistency with the legacy converters (`AbstractTokenConverter.baseAsset`, `SingleTokenConverter.baseAsset`, `RiskFundConverter.baseAsset`). It refers to the same thing there — the canonical token each destination holds, and the one every inbound token gets converted into. Happy to rename to `OUTPUT_ASSET` if you'd prefer a clean break from that naming.

File: contracts/ProtocolReserve/TokenBuyback.sol (hunk 2 above)
Comment (Debugger022) [reply contracts/ProtocolReserve/TokenBuyback.sol]: Yes — it's for Shortfall's per-pool accounting. `Shortfall.auction` → `RiskFundV2.transferReserveForAuction(comptroller, amount)` reads `poolAssetsFunds[comptroller][convertibleBaseAsset]` and reverts with `InsufficientPoolReserve` if the ledger isn't credited. Without the callback from `TokenBuyback`, the pool's bad debt can't be auctioned even though the tokens are in the contract. PLP/XVSVaultTreasury don't have that per-pool constraint — funds are pooled globally, no downstream reader — so no callback there.

File: contracts/ProtocolReserve/RiskFundV2.sol (hunk 3 above)
Comment (Debugger022) [reply contracts/ProtocolReserve/RiskFundV2.sol:224]: Agree with the direction for the steady state — if isolated pools are fully wound down and core pool doesn't auction via `Shortfall.sol`, then `poolAssetsFunds`, `transferReserveForAuction`, `sweepTokenFromPool`, and the shortfall setter are dead weight on BSC. Two concerns before pulling the trigger:

1. `RiskFundV2.shortfall` is still set on mainnet (`0xf37530A8a810Fcb501AA0Ecd0B0699388F0F2209`) and `Shortfall.sol` imports `IRiskFund` — removing those functions breaks that ABI contract. Would need a coordinated Shortfall deprecation or upgrade.

2. Residual `poolAssetsFunds[isolatedPool][asset]` entries from prior activity become unrecoverable without `sweepTokenFromPool`. Governance would need a sweep pass first.

Should we do it?

File: contracts/ProtocolReserve/TokenBuyback.sol (hunk 1 above)
Comment (fred-venus) [reply contracts/ProtocolReserve/TokenBuyback.sol:30]: then let's keep the convension

File: contracts/ProtocolReserve/TokenBuyback.sol (hunk 2 above)
Comment (fred-venus) [reply contracts/ProtocolReserve/TokenBuyback.sol]: got it, thx

File: contracts/ProtocolReserve/RiskFundV2.sol (hunk 3 above)
Comment (fred-venus) [reply contracts/ProtocolReserve/RiskFundV2.sol:224]: then how about keep the interface unchanged so that the bad debt liquidation flow still works but solely remove `poolAssetsFunds[isolatedPool][asset]` i.e. per pool accounting ? Reasons as follow:

1. we've already removed the corresponding logic in preSweepToken which means the accounting will no matter what break i.e. if ppl call sweepToken, the actual balance is gonna be less than accounting amt in  poolAssetsFunds
2. once we've removed `poolAssetsFunds`, the callback i.e. `updatePoolState` is no longer needed either, so we can further simplify new TokenBuyback contract, we dont need `if (isRiskFund) ... ` any more

Comment (Debugger022) [reply contracts/ProtocolReserve/RiskFundV2.sol:224]: Agree — cleaner than the full strip. Keeping the `IRiskFund` ABI intact preserves Shortfall compatibility, and on the `TokenBuyback` side dropping `IS_RISK_FUND` + `updatePoolState` + `ComptrollerRequired` makes all 7 instances homogeneous, which is a real simplification.

**Comment by fred-venus** [comment]
image


also, let's follow the income allocation here, i.e, for treasury we will need 6 buyback contract and for prime we need 2, also  2 more for riskFund and xvsStore

**Review by fred-venus — APPROVED** [review]

**Review by GitGuru7 — APPROVED** [review]

File: contracts/TokenConverter/RiskFundConverter.sol
Comment (fred-venus) [inline contracts/TokenConverter/RiskFundConverter.sol:1]: this is also deprecated and we won't use it right ? i mean new buyback contract for riskFund is now TokenBuyback

Comment (Debugger022) [reply contracts/TokenConverter/RiskFundConverter.sol:1]: yes.

**Comment by fred-venus** [comment]
lgtm, employing ai to do the final check

1. deployed addrs match latest source code
2. cap starts with 30k per day, slippage event -> $500
3. matching the one used in one-off helper i.e.

**Review by fred-venus — APPROVED** [review]
